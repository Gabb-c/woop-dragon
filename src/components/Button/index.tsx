import styles from './styles.module.scss';
interface ButtonProps extends React.ButtonHTMLAttributes<HTMLButtonElement> {
  label?: string;
  loading?: boolean;
  flavor?: 'delete' | 'save' | 'default';
}

export const Button: React.FC<ButtonProps> = ({
  label,
  flavor,
  loading,
  ...props
}): JSX.Element => {
  return (
    <button
      {...props}
      className={!flavor || flavor === 'default' ? styles.button : styles[`button-${flavor}`]}
    >
      {!loading ? label : 'loading...'}
    </button>
  );
};
