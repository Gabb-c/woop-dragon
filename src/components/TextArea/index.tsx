import styles from './styles.module.scss';

interface TextAreaProps extends React.TextareaHTMLAttributes<HTMLTextAreaElement> {
  label?: string;
  error?: boolean;
  errorText?: string;
}

export const TextArea: React.FC<TextAreaProps> = ({
  label,
  error,
  errorText,
  ...props
}): JSX.Element => {
  return (
    <div className={styles.container}>
      <span className={error ? styles.inputErrorLabel : styles.inputLabel}>{label}</span>
      <textarea {...props} data-testid="textarea-test" className={styles.input} />
      {error && <span className={styles.inputErrorText}>{errorText}</span>}
    </div>
  );
};
