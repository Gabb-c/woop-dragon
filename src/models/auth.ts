export interface LoginResponse {
  access_token: string;
  user: User;
}

export interface User {
  name: string;
  email: string;
}
