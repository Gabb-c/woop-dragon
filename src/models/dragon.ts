export interface Dragon {
  readonly id: string;
  readonly createdAt: string;
  name: string;
  type: string;
  histories: string;
}
