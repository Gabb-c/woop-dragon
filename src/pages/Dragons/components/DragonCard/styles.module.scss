.card {
  display: flex;
  flex-direction: column;
  align-items: flex-start;
  justify-content: center;

  width: 20em;
  min-height: fit-content;

  padding: 1.2em;

  gap: 1em;

  color: #fefefe;
  background-color: #36393f;
  box-shadow: rgb(0, 0, 77) 0px 5px 10px -1px;
  border-radius: 0.5em;

  animation-duration: 0.2s;
  animation-name: animate-fade;
  animation-delay: 0.2s;
  animation-fill-mode: backwards;

  transition: all 0.6s ease;

  &:hover {
    transform: translateY(-8px);
    box-shadow: rgb(0, 0, 77) 0px 20px 30px -10px;
  }

  h1 {
    width: 100%;
    padding-bottom: 0.6rem;
    padding-top: 0.6rem;
    overflow: hidden;
    text-overflow: ellipsis;
  }

  h5 {
    max-width: 100%;
    overflow: hidden;
    text-overflow: ellipsis;
  }
}

.fields {
  display: flex;
  gap: 0.3em;
}

.dragonType {
  padding: 0.6em;
  border-radius: 0.8em;
  border: 1px solid #fefefe;

  letter-spacing: 0.1rem;
}

.dragonName {
  color: #646cff;
}

.dragonBirth {
  letter-spacing: 0.2rem;
}

@keyframes animate-fade {
  0% {
    opacity: 0;
  }
  10% {
    opacity: 0.1;
  }
  20% {
    opacity: 0.2;
  }
  30% {
    opacity: 0.3;
  }
  40% {
    opacity: 0.4;
  }
  50% {
    opacity: 0.5;
  }
  60% {
    opacity: 0.6;
  }
  70% {
    opacity: 0.7;
  }
  80% {
    opacity: 0.8;
  }
  90% {
    opacity: 0.9;
  }
  100% {
    opacity: 1;
  }
}
