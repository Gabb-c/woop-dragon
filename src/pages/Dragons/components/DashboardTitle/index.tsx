import styles from './styles.module.scss';
import { Button } from '@components/Button';
import { useAuth } from '@contexts/AuthContext';
import { useNavigate } from 'react-router-dom';

export const DashboardHeader = () => {
  const { signOut } = useAuth();

  const navigate = useNavigate();

  return (
    <div className={styles.container}>
      <Button label="Logout" flavor="delete" onClick={() => signOut()} />
      <h1>Dragons Dashboard</h1>
      <Button label="Add" flavor="save" onClick={() => navigate('/dragons/form')} />
    </div>
  );
};
