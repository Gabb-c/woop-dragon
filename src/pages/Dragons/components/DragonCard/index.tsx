import { Dragon } from '@models/dragon';
import { format } from 'date-fns';
import { Button } from 'src/pages/Login/components/Button';
import styles from './styles.module.scss';

interface DragonCardProps {
  dragon: Dragon;
  onDelete: (id: string) => void;
  onEdit?: (id: string) => void;
}

export const DragonCard: React.FC<DragonCardProps> = ({ dragon, onDelete }): JSX.Element => {
  return (
    <div className={styles.card}>
      <h5 className={styles.dragonType}>{dragon.type.toUpperCase()}</h5>
      <h1 className={styles.dragonName}>{dragon.name}</h1>
      <h3 className={styles.dragonBirth}>{format(new Date(dragon.createdAt), 'MM/dd/yyyy')}</h3>
      <div className={styles.fields}>
        <Button label="Edit" />
        <Button label="Delete" onClick={() => onDelete(dragon.id)} />
      </div>
    </div>
  );
};
