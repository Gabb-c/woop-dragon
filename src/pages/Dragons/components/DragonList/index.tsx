import { useEffect, useState } from 'react';
import { PropagateLoader } from 'react-spinners';

import DragonService from '@services/dragon';

import { DragonCard } from '../DragonCard';
import { Dragon } from '@models/dragon';

import styles from './styles.module.scss';
import { DashboardHeader } from '../DashboardTitle';

export const DragonList: React.FC = (): JSX.Element => {
  const [list, setList] = useState<Dragon[]>([]);
  const [loading, setLoading] = useState(true);

  const loadDragons = async () => {
    setLoading(true);
    await DragonService.getAllDragons()
      .then((response) => {
        setList(response);
        return setLoading(false);
      })
      .catch((error) => console.error(error));
  };

  const handleDelete = (id: string) => {
    setLoading(true);
    DragonService.deleteDragonById(id)
      .then(() => {
        return loadDragons();
      })
      .catch((error) => console.error(error));
  };

  useEffect(() => {
    loadDragons().catch((error) => console.error(error));
  }, []);

  return (
    <div className={styles.container}>
      <DashboardHeader />
      {loading ? (
        <PropagateLoader color="white" />
      ) : (
        <div className={styles.lisCtontainer}>
          {list.map((d, i) => (
            <DragonCard onDelete={handleDelete} dragon={d} key={i} />
          ))}
        </div>
      )}
    </div>
  );
};
