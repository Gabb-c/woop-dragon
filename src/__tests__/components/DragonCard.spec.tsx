import React from 'react';
import userEvent from '@testing-library/user-event';
import { render, screen, waitFor } from '@testing-library/react';
import { DragonCard } from '../../pages/Dragons/components/DragonCard';
import { Dragon } from '../../models/dragon';
import '@testing-library/jest-dom';

describe('DragonCard', () => {
  it('test if it renders in the DOM', () => {
    const mockDragon: Dragon = {
      createdAt: new Date().toISOString(),
      histories: '',
      name: 'Ysera',
      id: '0',
      type: 'dream',
    };

    render(<DragonCard dragon={mockDragon} onDelete={jest.fn} onEdit={jest.fn} />);

    const resTest = screen.getByText(/ysera/i);

    expect(resTest).toBeInTheDocument();
  });

  it('test if edit button function is triggered', async () => {
    const mockDragon: Dragon = {
      createdAt: new Date().toISOString(),
      histories: '',
      name: 'Ysera',
      id: '0',
      type: 'dream',
    };

    const mockOnDelete = jest.fn();
    const mockOnEdit = jest.fn();

    const { container } = render(
      <DragonCard dragon={mockDragon} onDelete={mockOnDelete} onEdit={mockOnEdit} />
    );

    const button = container.querySelector(`button[name="edit-dragon"]`);
    await userEvent.click(button as Element);

    await waitFor(() => {
      expect(mockOnEdit.mock.calls.length).toEqual(1);
    });
  });

  it('test if delete button function is triggered', async () => {
    const mockDragon: Dragon = {
      createdAt: new Date().toISOString(),
      histories: '',
      name: 'Ysera',
      id: '0',
      type: 'dream',
    };

    const mockOnDelete = jest.fn();
    const mockOnEdit = jest.fn();

    const { container } = render(
      <DragonCard dragon={mockDragon} onDelete={mockOnDelete} onEdit={mockOnEdit} />
    );

    const button = container.querySelector(`button[name="delete-dragon"]`);
    await userEvent.click(button as Element);

    await waitFor(() => {
      expect(mockOnDelete.mock.calls.length).toEqual(1);
    });
  });

  it('should match snapshot', () => {
    const mockDragon: Dragon = {
      createdAt: new Date().toISOString(),
      histories: '',
      name: 'Ysera',
      id: '0',
      type: 'dream',
    };

    render(<DragonCard dragon={mockDragon} onDelete={jest.fn} onEdit={jest.fn} />);

    const resTest = screen.getByText(/ysera/i);

    expect(resTest).toMatchSnapshot();
  });
});
