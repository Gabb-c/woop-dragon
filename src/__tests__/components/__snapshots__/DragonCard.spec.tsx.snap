// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`DragonCard should match snapshot 1`] = `
<h1>
  Ysera
</h1>
`;
